File: study/lua/src/app/scene/view/team/TeamInstrumentIcon.lua
--
-- Author: Liangxu
-- Date: 2017-02-20 17:47:24
--
local TeamInstrumentIcon = class("TeamInstrumentIcon")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local DataConst = require("app.const.DataConst")
local LogicCheckHelper = require("app.utils.LogicCheckHelper")
local InstrumentConst = require("app.const.InstrumentConst")

function TeamInstrumentIcon:ctor(target)
	self._target = target
	self._isOpen = false --是否开启
	self._pos = nil
	self._treasureId = nil
	self._totalList = nil
	self._noWearList = nil

	self._imageBg = ccui.Helper:seekNodeByName(self._target, "ImageBg")
	self._imageLock = ccui.Helper:seekNodeByName(self._target, "ImageLock")
	self._textTip = ccui.Helper:seekNodeByName(self._target, "TextTip")
	self._spriteAdd = ccui.Helper:seekNodeByName(self._target, "SpriteAdd")
	self._fileNodeName = ccui.Helper:seekNodeByName(self._target, "FileNodeName")
	cc.bind(self._fileNodeName, "CommonInstrumentName")
	self._fileNodeCommon = ccui.Helper:seekNodeByName(self._target, "FileNodeCommon")
	cc.bind(self._fileNodeCommon, "CommonInstrumentIcon")
	self._imageRedPoint = ccui.Helper:seekNodeByName(self._target, "ImageRedPoint")
	self._imageArrow = ccui.Helper:seekNodeByName(self._target, "ImageArrow")
	self._panelTouch = ccui.Helper:seekNodeByName(self._target, "PanelTouch")
    self._panelTouch:addClickEventListenerEx(handler(self, self._onPanelTouch))
    self:_initUnlock()
end

function TeamInstrumentIcon:_initUI()
	self._imageLock:setVisible(false)
    self._textTip:setVisible(false)
    self._spriteAdd:setVisible(false)
	self._fileNodeCommon:setVisible(false)
	self._fileNodeName:setVisible(false)
	self._imageRedPoint:setVisible(false)
	self._imageArrow:setVisible(false)
end

function TeamInstrumentIcon:_initUnlock( ... )
    -- body
    local isOpen, __, info = LogicCheckHelper.funcIsOpened(FunctionConst.FUNC_INSTRUMENT)
    if not isOpen then
        self._textTip:setPosition(cc.p(36, 42))
        self._textTip:setVisible(true)
        self._textTip:setFontSize(14)
		self._textTip:setString(Lang.get("instrument_txt_open", {level = info.level}))
	end 
end

function TeamInstrumentIcon:updateIcon(pos, heroBaseId)
	self:_initUI()
	local isOpen, des, info = LogicCheckHelper.funcIsOpened(FunctionConst.FUNC_INSTRUMENT)
	self._isOpen = isOpen
    if not isOpen then
        
		self._imageLock:setVisible(true)
        self._textTip:setVisible(true)
        self._textTip:setPosition(cc.p(36, 42))
        self._textTip:setFontSize(14)
		self._textTip:setString(Lang.get("instrument_txt_open", {level = info.level}))
		return
	end 

	self._pos = pos
	self._heroBaseId = heroBaseId
	self._instrumentId = G_UserData:getBattleResource():getResourceId(pos, 3, 1)
	if self._instrumentId then
		local data = G_UserData:getInstrument():getInstrumentDataWithId(self._instrumentId)
		local baseId = data:getBase_id()
		local limitLevel = data:getLimit_level()
		local param = TypeConvertHelper.convert(TypeConvertHelper.TYPE_INSTRUMENT, baseId, nil, nil, limitLevel)
		self._fileNodeCommon:setVisible(true)
		self._fileNodeCommon:updateUI(baseId, nil, limitLevel)
		-- self._fileNodeCommon:loadColorBg(param.icon_bg2)
		self._fileNodeCommon:setRlevel(data:getLevel())
		self._fileNodeName:setVisible(true)
		self._fileNodeName:setName(baseId, nil, limitLevel)
		self._fileNodeCommon:hideBg()
	else
		self._totalList, self._noWearList = G_UserData:getInstrument():getReplaceInstrumentListWithSlot(self._pos, self._heroBaseId)
		if #self._noWearList > 0 then
            self._spriteAdd:setVisible(true)
            self._textTip:setPosition(cc.p(36, -9))
            self._textTip:setVisible(true)
            self._textTip:setFontSize(18)
			self._textTip:setString(Lang.get("instrument_add_tip"))
			local UIActionHelper = require("app.utils.UIActionHelper")
			UIActionHelper.playBlinkEffect(self._spriteAdd)
		end
	end
end

function TeamInstrumentIcon:_onPanelTouch()
	if not self._isOpen then
		return
	end
	
	if self._instrumentId then
		G_SceneManager:showScene("instrumentDetail", self._instrumentId, InstrumentConst.INSTRUMENT_RANGE_TYPE_2)
	else
		if #self._noWearList == 0 then
			local PopupItemGuider = require("app.ui.PopupItemGuider").new(Lang.get("way_type_get"))
			local configInfo = require("app.config.hero").get(self._heroBaseId)
			assert(configInfo, string.format("hero config can not find id = %d", self._heroBaseId))
			local instrumentId = configInfo.instrument_id
			if instrumentId > 0 then
				PopupItemGuider:updateUI(TypeConvertHelper.TYPE_INSTRUMENT, instrumentId)
		    	PopupItemGuider:openWithAction()
			end
		else
			local PopupChooseInstrumentHelper = require("app.ui.PopupChooseInstrumentHelper")
			local popup = require("app.ui.PopupChooseInstrument").new()
			local callBack = handler(self, self._onChooseInstrument)
			popup:setTitle(Lang.get("instrument_wear_title"))
			popup:updateUI(PopupChooseInstrumentHelper.FROM_TYPE1, callBack, self._totalList)
			popup:openWithAction()
		end
	end
end

function TeamInstrumentIcon:_onChooseInstrument(instrumentId)
	G_UserData:getInstrument():c2sAddFightInstrument(self._pos, instrumentId)
end

function TeamInstrumentIcon:showRedPoint(visible)
	self._imageRedPoint:setVisible(visible)
end

function TeamInstrumentIcon:showUpArrow(visible)
	local UIActionHelper = require("app.utils.UIActionHelper")
	self._imageArrow:setVisible(visible)
	if visible then
		UIActionHelper.playFloatEffect(self._imageArrow)
	end
end

function TeamInstrumentIcon:onlyShow(data)
	self:_initUI()
	self._panelTouch:setEnabled(false)
	if data then
		local baseId = data:getBase_id()
		local limitLevel = data:getLimit_level()
		local param = TypeConvertHelper.convert(TypeConvertHelper.TYPE_INSTRUMENT, baseId, nil, nil, limitLevel)
		self._fileNodeCommon:setVisible(true)
		self._fileNodeCommon:updateUI(baseId, nil, limitLevel)
		-- self._fileNodeCommon:loadColorBg(param.icon_bg2)
		self._fileNodeCommon:setRlevel(data:getLevel())
		self._fileNodeName:setVisible(true)
		self._fileNodeCommon:hideBg()
		self._fileNodeName:setName(baseId, nil, limitLevel)
	end
end

function TeamInstrumentIcon:setVisible(visible)
	self._target:setVisible(visible)
end

function TeamInstrumentIcon:showTextBg(bShow)
	self._fileNodeName:showTextBg(bShow)
	self._fileNodeName:setFontSize(18)
end

function TeamInstrumentIcon:showUnlockView(isVisible)
    -- body
    self._imageLock:setVisible(not isVisible)
    self._spriteAdd:setVisible(not isVisible)
    self._fileNodeCommon:setVisible(not isVisible)
    self._imageRedPoint:setVisible(not isVisible)
    self._fileNodeName:setVisible(not isVisible)
    self._imageArrow:setVisible(not isVisible)
end

return TeamInstrumentIcon

File: study/ts/script/scene/view/team/TeamInstrumentIcon.ts
import CommonInstrumentName from "../../../ui/component/CommonInstrumentName";
import CommonInstrumentIcon from "../../../ui/component/CommonInstrumentIcon";
import { handler } from "../../../utils/handler";
import { FunctionConst } from "../../../const/FunctionConst";
import { FunctionCheck } from "../../../utils/logic/FunctionCheck";
import { Lang } from "../../../lang/Lang";
import { G_UserData, G_SceneManager, G_ConfigLoader } from "../../../init";
import { TypeConvertHelper } from "../../../utils/TypeConvertHelper";
import InstrumentConst from "../../../const/InstrumentConst";
import { assert } from "../../../utils/GlobleFunc";
import UIActionHelper from "../../../utils/UIActionHelper";
import PopupItemGuider from "../../../ui/PopupItemGuider";
import { UIPopupHelper } from "../../../utils/UIPopupHelper";
import { ConfigNameConst } from "../../../const/ConfigNameConst";
import PopupChooseInstrument from "../../../ui/popup/PopupChooseInstrument";
import PopupChooseInstrumentHelper from "../../../ui/popup/PopupChooseInstrumentHelper";
import UIHelper from "../../../utils/UIHelper";
import { LogicCheckHelper } from "../../../utils/LogicCheckHelper";

const {ccclass, property} = cc._decorator;

@ccclass
export default class TeamInstrumentIcon extends cc.Component {

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBg: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageLock: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textTip: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _spriteAdd: cc.Sprite = null;

    @property({
        type: CommonInstrumentName,
        visible: true
    })
    _fileNodeName: CommonInstrumentName = null;

    @property({
        type: CommonInstrumentIcon,
        visible: true
    })
    _fileNodeCommon: CommonInstrumentIcon = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageRedPoint: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageArrow: cc.Sprite = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _panelTouch: cc.Node = null;

    private _isOpen:boolean;
    private _pos:number;
    private _heroBaseId:number;
    private _instrumentId:number;
    private _treasureId:number;
    private _totalList:Array<any>;
    private _noWearList:Array<any>;
    _isOnlyShow: boolean;


    onLoad() {
        this._isOpen = false;
        this._pos = null;
        this._treasureId = null;
        this._totalList = null;
        this._noWearList = null;
        this._panelTouch.on(cc.Node.EventType.TOUCH_END,handler(this, this._onPanelTouch));
        UIHelper.enableOutline(this._textTip,new cc.Color(119,41,9),2)
        this._initUnlock();
    }
    _initUI() {
        this._imageLock.node.active = (false);
        this._textTip.node.active = (false);
        this._spriteAdd.node.active = (false);
        this._fileNodeCommon.node.active = (false);
        this._fileNodeName.node.active = (false);
        this._imageRedPoint.node.active = (false);
        this._imageArrow.node.active = (false);
    }
    _initUnlock (){
        var [isOpen, __, info] = LogicCheckHelper.funcIsOpened(FunctionConst.FUNC_INSTRUMENT);
        if (!isOpen) {
            this._textTip.fontSize = (14);
            this._textTip.node.setPosition(0,0);
            this._textTip.node.active = (true);
            this._textTip.string = (Lang.get('instrument_txt_open', { level: info.level }));
        }
    }
    updateIcon(pos, heroBaseId) {
        this._isOnlyShow = false;
        this._initUI();
        var ret = FunctionCheck.funcIsOpened(FunctionConst.FUNC_INSTRUMENT);
        var isOpen = ret[0];
        var des = ret[1]; 
        var info = ret[2];
        this._isOpen = isOpen;
        if (!isOpen) {
            this._imageLock.node.active = (true);
            this._textTip.node.active = (true);
            this._textTip.node.setPosition(cc.v2(36, 42));
            this._textTip.fontSize = (14);
            this._textTip.string = (Lang.get('instrument_txt_open', { level: info.level }));
            return;
        }
        this._pos = pos;
        this._heroBaseId = heroBaseId;
        this._instrumentId = G_UserData.getBattleResource().getResourceId(pos, 3, 1);
        if (this._instrumentId) {
            var data = G_UserData.getInstrument().getInstrumentDataWithId(this._instrumentId);
            var baseId = data.getBase_id();
            var limitLevel = data.getLimit_level();
            var param = TypeConvertHelper.convert(TypeConvertHelper.TYPE_INSTRUMENT, baseId, null, null, limitLevel);
            this._fileNodeCommon.node.active = (true);
            this._fileNodeCommon.updateUI(baseId, null, limitLevel);
            this._fileNodeCommon.setRlevel(data.getLevel());
            this._fileNodeName.node.active = (true);
            this._fileNodeName.setName(baseId, null, limitLevel);
            this._fileNodeCommon.hideBg();
        } else {
            var ret:any[] = G_UserData.getInstrument().getReplaceInstrumentListWithSlot(this._pos, this._heroBaseId);
            this._totalList = ret[0], 
            this._noWearList = ret[1];
            if (this._noWearList&&this._noWearList.length > 0) {
                this._spriteAdd.node.active = (true);
                this._textTip.node.active = (true);
                this._textTip.node.setPosition(cc.v2(36, -9));
                this._textTip.fontSize = (18);
                this._textTip.node.setPosition(0,-51);
                this._textTip.string = (Lang.get('instrument_add_tip'));
                UIActionHelper.playBlinkEffect(this._spriteAdd.node);
            }
        }
    }
    _onPanelTouch() {
        if (!this._isOpen) {
            return;
        }
        if (this._instrumentId) {
            G_SceneManager.showScene('instrumentDetail', this._instrumentId, InstrumentConst.INSTRUMENT_RANGE_TYPE_2);
        } else {
            if (this._noWearList.length == 0) {
                var configInfo = G_ConfigLoader.getConfig(ConfigNameConst.HERO).get(this._heroBaseId);
              //assert((configInfo, cc.js.formatStr('hero config can not find id = %d', this._heroBaseId));
                var instrumentId = configInfo.instrument_id;
                if (instrumentId > 0) {
                    UIPopupHelper.popupItemGuider(function(popupItemGuider:PopupItemGuider){
                        popupItemGuider.updateUI(TypeConvertHelper.TYPE_INSTRUMENT, instrumentId);
                        popupItemGuider.setTitle(Lang.get('way_type_get'));
                    }.bind(this))
                }
            } else {
                var callBack = handler(this, this._onChooseInstrument);
                var totalList = this._totalList;
                UIPopupHelper.popupChooseInstrument(function(popup:PopupChooseInstrument){
                    popup.openWithAction();
                    popup.setTitle(Lang.get('instrument_wear_title'));
                    popup.updateUI(PopupChooseInstrumentHelper.FROM_TYPE1, callBack, totalList);
                });
            }
        }
    }
    _onChooseInstrument(instrumentId) {
        G_UserData.getInstrument().c2sAddFightInstrument(this._pos, instrumentId);
    }
    showRedPoint(visible) {
        this._imageRedPoint.node.active = (visible);
    }
    showUpArrow(visible) {
        this._imageArrow.node.active = (visible);
        if (visible) {
            UIActionHelper.playFloatEffect(this._imageArrow.node);
        }
    }
    onlyShow(data) {
        this._initUI();
        //this._panelTouch.setEnabled(false);
        if (data) {
            var baseId = data.getBase_id();
            var limitLevel = data.getLimit_level();
            var param = TypeConvertHelper.convert(TypeConvertHelper.TYPE_INSTRUMENT, baseId, null, null, limitLevel);
            this._fileNodeCommon.node.active = (true);
            this._fileNodeCommon.updateUI(baseId, null, limitLevel);
            this._fileNodeCommon.setRlevel(data.getLevel());
            this._fileNodeName.node.active = (true);
            this._fileNodeCommon.hideBg();
            this._fileNodeName.setName(baseId, null, limitLevel);
        }
    }
    setVisible(visible) {
        this.node.active = (visible);
    }
    showTextBg(bShow) {
        this._fileNodeName.showTextBg(bShow);
        this._fileNodeName.setFontSize(18);
    }

    showUnlockView(isVisible) {
        this._imageLock.node.active = (!isVisible);
        this._spriteAdd.node.active = (!isVisible);
        this._fileNodeCommon.node.active = (!isVisible);
        this._imageRedPoint.node.active = (!isVisible);
        this._fileNodeName.node.active = (!isVisible);
        this._imageArrow.node.active = (!isVisible);
    }
}